glommio/src/io/dma_open_options.rs

```diff
@@ -0,0 +1,192 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the
+// MIT/Apache-2.0 License, at your convenience
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2020 Datadog, Inc.
+//
+use crate::io::dma_file::DmaFile;
+use std::io;
+use std::path::Path;
+
+/// Options and flags which can be used to configure how a file is opened.
+///
+/// This builder exposes the ability to configure how a [`DmaFile`] is opened
+/// and what operations are permitted on the open file.
+/// The [`DmaFile::open`] and [`DmaFile::create`] methods are aliases for commonly used options using this builder.
+#[derive(Clone, Debug)]
+pub struct DmaOpenOptions {
+    read: bool,
+    pub(super) write: bool,
+    truncate: bool,
+    create: bool,
+    create_new: bool,
+    // system-specific
+    pub(super) custom_flags: libc::c_int,
+    pub(super) mode: libc::mode_t,
+}
+
+impl Default for DmaOpenOptions {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DmaOpenOptions {
+    /// Creates a blank new set of options ready for configuration.
+    ///
+    /// All options are initially set to false.
+    pub fn new() -> Self {
+        Self {
+            read: false,
+            write: false,
+            truncate: false,
+            create: false,
+            create_new: false,
+            // system-specific
+            custom_flags: 0,
+            // previously, we defaulted to 0o644, but 0o666 is usedd by libstd
+            mode: 0o666,
+        }
+    }
+
+    // the following methods are directly taken from std/sys/unix/fs.rs
+    // NOTE(zserik): I omitted `append` from the following methods.
+    //               I don't think it makes much sense for unbuffered files.
+
+    /// Sets the option for read access.
+    ///
+    /// This option, when true, will indicate that the file should be read-able if opened.
+    pub fn read(&mut self, read: bool) -> &mut Self {
```

Comment:
shouldn't those be `mut self` returning `Self` instead ? (consuming self)

Reply:
No, the pattern used seems correct, and is directly copied from `libstd`, so it's correct, because the actual `open` fn takes a `&self`, which makes this work. It has the benefit that it allows both classic chaining and simple method calling, which is good for usability.